### PR TITLE
feat(medusa): Product collections, types and tags

### PR DESCRIPTION
**Pull request by olivermrbl**
Adds product collection, tag, and type to allow for more granular differentiation

Adds autogenerated product handle, if this is not provided. Will use a kebab-cased title

Adds product collection, tag, and type support to Contentful plugin

Adds product integration tests, that tests product collection, tag, and type

### CODE REVIEW

**Comment by olivermrbl** [comment]
Marking this as WIP, as I want to add integration tests when pull request #163 is merged in :)

File: packages/medusa/src/api/routes/admin/products/index.js [hunk 1]
```diff
@@ -8,6 +8,11 @@ export default app => {
 
   route.post("/", middlewares.wrap(require("./create-product").default))
   route.post("/:id", middlewares.wrap(require("./update-product").default))
+  route.get("/types", middlewares.wrap(require("./list-types").default))
+  route.get(
+    "/tag-usage",
+    middlewares.wrap(require("./list-tag-usage-count").default)
+  )
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/products/index.js:15]: Think we should be able to do something along these lines instead: 
```
GET /tags?sort=usage_count
```

If this is too complicated to implement (could imagine it requires some subqueries in `tagRepo.find`) let's just remove it until we find a better pattern for querying with subqueries.

File: packages/medusa/src/migrations/1612359285299-product_option_product_id.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      `COMMENT ON COLUMN "product_option"."product_id" IS NULL`
+    )
+    await queryRunner.query(
+      `ALTER TABLE "product_option" ALTER COLUMN "product_id" DROP NOT NULL`
+    )
+    await queryRunner.query(
+      `ALTER TABLE "product_option" ADD CONSTRAINT "FK_e634fca34f6b594b87fdbee95f6" FOREIGN KEY ("product_id") REFERENCES "product"("id") ON DELETE NO ACTION ON UPDATE NO ACTION`
+    )
+  }
+}
```
Comment (srindom) [inline packages/medusa/src/migrations/1612359285299-product_option_product_id.ts:29]: Looks like this removes the foreign key constraint i.e. referential integrity is broken. Not sure why it generates it like that, but I think the current models will work without this as the `product_option` table already has foreign key with the correct name: `product_id`

**Review by srindom — CHANGES_REQUESTED** [review]
Epic

File: packages/medusa/src/migrations/1612359285299-product_option_product_id.ts (hunk 2 above)
Comment (olivermrbl) [reply packages/medusa/src/migrations/1612359285299-product_option_product_id.ts:29]: Yeah to be honest.. This migration is looking really weird. We really only need `ALTER TABLE "product_option" ALTER COLUMN "product_id" SET NOT NULL` as far as I can see. Correct me if I am wrong? @srindom

File: packages/medusa/src/api/routes/admin/products/index.js (hunk 1 above)
Comment (olivermrbl) [reply packages/medusa/src/api/routes/admin/products/index.js:15]: The usage count query was tricky enough, so I say, let's wait till we need tag listing in more ways than now.

File: packages/medusa/src/migrations/1612359285299-product_option_product_id.ts (hunk 2 above)
Comment (srindom) [reply packages/medusa/src/migrations/1612359285299-product_option_product_id.ts:29]: I believe TypeORM adds a `NOT NULL` constraint by default to the foreign keys created. So the migration is unnecessary altogether.

Comment (olivermrbl) [reply packages/medusa/src/migrations/1612359285299-product_option_product_id.ts:29]: But the column `product_id` is missing. Before this migration, we were not able to do `productOptionRepo.create({ ...someOption, product_id: product.id })`

**Pull request closed without merging**